include/libopencm3/stm32/f7/flash.h

```diff
@@ -28,10 +32,174 @@
  * along with this library.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#ifndef LIBOPENCM3_FLASH_H
-#define LIBOPENCM3_FLASH_H
+/*
+ * For details see:
+ * PM0081 Programming manual: STM32F40xxx and STM32F41xxx Flash programming
+ * September 2011, Doc ID 018520 Rev 1
+ * https://github.com/libopencm3/libopencm3-archive/blob/master/st_micro/DM00023388.pdf
+ */
+
+/*
+ * Differences between F7 and F4:
+ * 1. 	icache and dcache are now combined into a unified ART cache. The CPU has
+ * 	its own d/i-caches, but those are unrelated to this. They are on the
+ *	AXIM bus.
+ * 2. 	There's an OPTCR1 is now used for boot addresses. Write protect bits
+ * 	are in OPTCR. Why does F4 have 2 copies of nWRP?
+ * 3. 	Latency field in FLASH_ACR is now 4 bits. Some CPU frequencies supported
+ * 	by F7 require more than 7 wait states.
+ * 4. 	FLASH_SR_PGSERR (programming sequence error) is now FLASH_SR_ERSERR (
+ *	erase sequence error).
+ * 5. 	FLASH_SR_BSY field is now read-only. Seems to also be read-only in F4?
+ *	Why did we have a clear busy flag function?
+ * 6. 	There are now two watchdogs - IWDG (independent watchdog) and WWDG (
+ * 	window watchdog).
+ */
+
+/**@{*/
+/* --- FLASH registers ----------------------------------------------------- */
+
+#define FLASH_ACR			MMIO32(FLASH_MEM_INTERFACE_BASE + 0x00)
+#define FLASH_KEYR			MMIO32(FLASH_MEM_INTERFACE_BASE + 0x04)
+#define FLASH_OPTKEYR			MMIO32(FLASH_MEM_INTERFACE_BASE + 0x08)
+#define FLASH_SR			MMIO32(FLASH_MEM_INTERFACE_BASE + 0x0C)
+#define FLASH_CR			MMIO32(FLASH_MEM_INTERFACE_BASE + 0x10)
+#define FLASH_OPTCR			MMIO32(FLASH_MEM_INTERFACE_BASE + 0x14)
+#define FLASH_OPTCR1			MMIO32(FLASH_MEM_INTERFACE_BASE + 0x18)
+
+/* --- FLASH Keys -----------------------------------------------------------*/
+
+#define FLASH_KEYR_KEY1			((uint32_t)0x45670123)
+#define FLASH_KEYR_KEY2			((uint32_t)0xcdef89ab)
+#define FLASH_OPTKEYR_KEY1		((uint32_t)0x08192a3b)
+#define FLASH_OPTKEYR_KEY2		((uint32_t)0x4c5d6e7f)
```

Comment:
Is the cast here necessary?  shouldn't they just get a UL suffix to be the correct literal instead?